File: project/module-core/src/main/java/cn/fd/ratziel/core/contextual/ArgumentContext.java
package cn.fd.ratziel.core.contextual;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.function.Supplier;

/**
 * ArgumentContext - 参数上下文
 * 本质为一个参数容器
 *
 * @author TheFloodDragon
 * @since 2024/7/16 11:56
 */
public interface ArgumentContext {

    /**
     * 弹出指定类型的参数
     *
     * @throws ArgumentNotFoundException 当无法找到指定类型的参数时抛出
     */
    <T> @NotNull T pop(@NotNull Class<@NotNull T> type) throws ArgumentNotFoundException;

    /**
     * 弹出指定类型的参数
     * 若无法找到, 则返回空
     */
    <T> @Nullable T popOrNull(@NotNull Class<@NotNull T> type);

    /**
     * 弹出指定类型的参数
     * 若无法找到, 则返回默认值
     */
    default <T> @NotNull T popOr(@NotNull Class<T> type, @NotNull Supplier<@NotNull T> def) {
        final T obj = this.popOrNull(type);
        return obj == null ? def.get() : obj;
    }

    /**
     * 弹出指定类型的参数
     * 若无法找到, 则添加默认值
     */
    default <T> @NotNull T popOrPut(@NotNull Class<T> type, @NotNull Supplier<@NotNull T> def) {
        T obj = this.popOrNull(type);
        if (obj == null) {
            obj = def.get();
            this.put(obj);
        }
        return obj;
    }

    /**
     * 添加一个参数
     */
    void put(@NotNull Object element);

    /**
     * 添加多个参数
     */
    default void putAll(@NotNull Iterable<@NotNull Object> elements) {
        for (Object element : elements) put(element);
    }

    /**
     * 删除一个参数
     */
    void remove(@NotNull Class<?> type);

    /**
     * 获取所有参数
     */
    @NotNull
    Iterable<@NotNull Object> args();

}

File: project/module-core/src/main/kotlin/cn/fd/ratziel/core/contextual/ArgumentContext.kt
package cn.fd.ratziel.core.contextual


/**
 * ArgumentContext - 参数上下文 (本质为一个参数容器)
 *
 * @author TheFloodDragon
 * @since 2024/7/16 11:56
 */
interface ArgumentContext {

    /**
     * 弹出指定类型的参数
     *
     * @throws ArgumentNotFoundException 当无法找到指定类型的参数时抛出
     */
    @Throws(ArgumentNotFoundException::class)
    fun <T : Any> pop(type: Class<T>): T {
        return popOrNull(type) ?: throw ArgumentNotFoundException(type)
    }

    /**
     * 弹出指定类型的参数
     * 若无法找到, 则返回空
     */
    fun <T : Any> popOrNull(type: Class<T>): T?

    /**
     * 弹出指定类型的参数
     * 若无法找到, 则返回默认值
     */
    fun <T : Any> popOr(type: Class<T>, def: () -> T): T {
        val obj = this.popOrNull<T>(type)
        return obj ?: def()
    }

    /**
     * 弹出指定类型的参数
     * 若无法找到, 则添加默认值
     */
    fun <T : Any> popOrPut(type: Class<T>, def: () -> T): T {
        var obj = this.popOrNull<T>(type)
        if (obj == null) {
            obj = def()
            this.put(obj)
        }
        return obj
    }

    /**
     * 添加一个参数
     */
    fun put(element: Any)

    /**
     * 添加多个参数
     */
    fun putAll(elements: Iterable<Any>) {
        for (element in elements) put(element)
    }

    /**
     * 删除一个参数
     */
    fun remove(type: Class<*>)

    /**
     * 复制一份新的参数上下文
     */
    fun copy(): ArgumentContext

    /**
     * 获取所有参数
     */
    fun args(): Iterable<Any>

}